[subelements lookup] allow subelement property to be nested and add optional 'skip_missing' flag

srgvg opened the pull request:
Allow the subelement property to the subelements lookup plugin to be nested:

```

---
users:
  - user: bob
    password: system-password
    mysql:
      password: mysql-password
      hosts:
        - "%"
        - "{{ ansible_hostname }}"
        - "127.0.0.1"
        - "::1"
        - "localhost"
      privs:
        - "*.*:SELECT"
        - "DB1.*:ALL"
```

you can now refer to the mysql.hosts subkey 

```
- name: Setup MySQL users
  mysql_user: name={{ item.0.user }} password={{ item.0.mysql.password }} host={{ item.1 }} priv={{ item.0.mysql.privs | join('/') }}
  with_subelements:
    - users
    - mysql.hosts
```

closes #9863 

A second set of patches addresses and closes #9827 by implementing an optional third list item, holding a dictionary of flags, allowing 'skip_missing: yes' to be set, which will let the plugin just skip list items that do not contain the given subkey, instead of raising an error.

srgvg committed:
Allow subelement property to be nested

The subelement property can now refer to a deeper nested key.
closes #9863

srgvg committed:
Documentation for nested subelements

srgvg committed:
Integration tests for nested subelements

wolph commented:
Just a thought, why use a `.` syntax instead of just more subelement items?

Making it go recursive like this as well might be more flexible:

``` yaml
- name: Setup MySQL users
  mysql_user: name={{ item.0.user }} password={{ item.0.mysql.password }} host={{ item.1 }} priv={{ item.0.mysql.privs | join('/') }}
  with_subelements:
    - users
    - mysql
    - hosts
```

Just in case of a list of hashes.

srgvg commented:
@WoLpH for starters, it seems to me the dot notation is more in line with the jinja2 notation.
Also, I just bumped into #9827 where we propose to add a parameter to just skip list items that don't have the subkey.

wolph commented:
True, instead is actually the wrong wording. I was thinking of a case like this:

``` yaml
users:
  - user: bob
    password: system-password
    mysql:
      password: mysql-password
      hosts:
        production:
          -"%"
          - "{{ ansible_hostname }}"
        local:
          - "127.0.0.1"
          - "::1"
          - "localhost"
```

srgvg commented:
@WoLpH not sure how your last example is any different? 

```
with_subelements:
- users
- mysql.hosts.production
```

would work with my patch.

srgvg committed:
Add optional 'skip_missing' flag to subelements

plugin. If set to true, the plugin will just skip list items that miss the
given subkey, instead of raising an error.

closes #9827

srgvg committed:
Documentation for subelements 'skip_missing' flag

srgvg committed:
Tests for optional subelements 'skip_missing' flag

wolph commented:
@srvg: Would your current code work if you want to iterate through all available groups? So both local and production? It's probably a remote use-case but it might still be useful :)

srgvg commented:
No, it wouldn't. you could only point the key to mysql.hosts.production OR mysql.hosts.local.

There should be a more generic way to combine lookup plugins. I proposed a patch for that two years ago (#4627) but I never liked the syntax I had to implement. This should be something to look into post v2.

This being said, my patch here is on the v1 tree, and I heard v2 changes some things in lookup plugins, so I'll need to update this patch for v2.

wolph commented:
If you need any help with the patches, let me know :)
I'd be happy to do some work as well.

As for the syntax... how about something like this:
1. Every line works through the parent list/dict
2. Every dot is just immediate access (i.e. `__getitem__`)

Still won't cover all cases but it should allow for slightly more complex examples.

srgvg commented:
Closing this in favour of the one update to v2: #10995
